ast/compile.go

```diff
@@ -4762,7 +4762,18 @@ func checkUnusedAssignedAndArgVars(loc *Location, stack *localDeclaredVars, used
 	unused = unused.Diff(rewrittenUsed)
 
 	for _, gv := range unused.Sorted() {
-		errs = append(errs, NewError(CompileErr, loc, "assigned var %v unused", dvs.reverse[gv]))
+		switch dvs.occurrence[dvs.reverse[gv]] {
+		case assignedVar:
+			for i := range body {
+				if body[i].Vars(VarVisitorParams{}).Contains(gv) {
+					// TODO: better way to get location of assignment?
+					errs = append(errs, NewError(CompileErr, body[i].Loc(), "assigned var %v unused", dvs.reverse[gv]))
+				}
+			}
+
+		case argVar:
```

Comment:
We have some [changes](https://github.com/open-policy-agent/opa/pull/5760) coming shortly that'll make this check redundant here. Once that's on `main`, and merged into here, we can drop this `switch/case`, and assume all unused vars are assignments.